openquake/hazardlib/gsim/gsim_table.py

```diff
@@ -0,0 +1,545 @@
+            # Interpolate magnitude - linear-log space
+            mag_interpolator = interp1d(self.magnitudes, period_table, axis=1)
+            output_table = 10.0 ** mag_interpolator(rctx.mag)
+        return output_table
+
+    def get_sigma_tables(self, imt, rctx, stddev_types):
+        """
+        Returns modification factors for the standard deviations, given the
+        rupture and intensity measure type.
+
+        :returns:
+            List of standard deviation modification tables, each as an array
+            of [Number Distances, Number Levels]
+
+        """
+        output_tables = []
+        for stddev_type in stddev_types:
+            # For PGA and PGV only needs to apply magnitude interpolation
+            if isinstance(imt, (imt_module.PGA, imt_module.PGV)):
+                interpolator = interp1d(self.magnitudes,
+                                        self.sigma[stddev_type][str(imt)],
+                                        axis=2)
+                output_tables.append(
+                    interpolator(rctx.mag).reshape(self.shape[0],
+                                                   self.shape[3]))
+
+            else:
+                # For spectral accelerations - need two step process
+                # Interpolate period
+                interpolator = interp1d(numpy.log10(self.periods),
+                                        self.sigma[stddev_type]["SA"],
+                                        axis=1)
+                period_table = interpolator(numpy.log10(imt.period))
+                mag_interpolator = interp1d(self.magnitudes,
+                                            period_table,
+                                            axis=1)
+                output_tables.append(mag_interpolator(rctx.mag))
+        return output_tables
+
+
+class GMPETable(GMPE):
+    """
+    Implements ground motion prediction equations in the form of a table from
+    which the expected ground motion intensity levels and standard deviations
+    are interpolated.
+
+    In a GMPE tables the expected ground motions for each of the IMTs over the
+    range of magnitudes and distances are stored in an hdf5 file on the path
+    specified by the user.
+
+    In this version of the GMPE the expected values are interpolated to the
+    required IMT, magnitude and distance in three stages.
+
+    i) Initially the correct IMT values are identified, interpolating in
+       log-T|log-IML space between neighbouring spectral periods.
+
+    ii) The IML values are then interpolated to the correct magnitude using
+        linear-M|log-IML space
+
+    iii) The IML values are then interpolated to the correct distance via
+         linear-D|linear-IML interpolation
+    """
+    DEFINED_FOR_TECTONIC_REGION_TYPE = ""
+
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = set(())
+
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = ""
+
+    DEFINED_FOR_STANDARD_DEVIATION_TYPES = set((const.StdDev.TOTAL,))
```

Comment:
If sigma_tot is not available then it won't be possible to use the GMPE in a hazard curve calculation, so I have to assume that it is provided. I can raise an explicit error somewhere if it is not provided, but probably this will be caught somewhere else in the calculation.